xls/modules/zstd/buffer.x

```diff
@@ -0,0 +1,451 @@
+
+// Structure for returning Buffer and BufferStatus together
+pub struct BufferResult<BSIZE: u32> {
+    buffer: Buffer<BSIZE>,
+    status: BufferStatus
+}
+
+// Checks whether a `buffer` can fit `data`
+pub fn buffer_can_fit<BSIZE: u32, DSIZE: u32>(buffer: Buffer<BSIZE>, data: bits[DSIZE]) -> bool {
+    buffer.length + DSIZE <= BSIZE
+}
+
+#[test]
+fn test_buffer_can_fit() {
+    let buffer = Buffer<u32:32> { contents: u32:0, length: u32:0 };
+    assert_eq(buffer_can_fit(buffer, bits[0]:0), true);
+    assert_eq(buffer_can_fit(buffer, u16:0), true);
+    assert_eq(buffer_can_fit(buffer, u32:0), true);
+    assert_eq(buffer_can_fit(buffer, u33:0), false);
+
+    let buffer = Buffer<u32:32> { contents: u32:0, length: u32:16 };
+    assert_eq(buffer_can_fit(buffer, bits[0]:0), true);
+    assert_eq(buffer_can_fit(buffer, u16:0), true);
+    assert_eq(buffer_can_fit(buffer, u17:0), false);
+    assert_eq(buffer_can_fit(buffer, u32:0), false);
+
+    let buffer = Buffer<u32:32> { contents: u32:0, length: u32:32 };
+    assert_eq(buffer_can_fit(buffer, bits[0]:0), true);
+    assert_eq(buffer_can_fit(buffer, u1:0), false);
+    assert_eq(buffer_can_fit(buffer, u16:0), false);
+    assert_eq(buffer_can_fit(buffer, u32:0), false);
+}
+
+// Checks whether a `buffer` has at least `length` amount of data
+pub fn buffer_has_at_least<BSIZE: u32>(buffer: Buffer<BSIZE>, length: u32) -> bool {
+    length <= buffer.length
+}
+
+#[test]
+fn test_buffer_has_at_least() {
+    let buffer = Buffer { contents: u32:0, length: u32:0 };
+    assert_eq(buffer_has_at_least(buffer, u32:0), true);
+    assert_eq(buffer_has_at_least(buffer, u32:16), false);
+    assert_eq(buffer_has_at_least(buffer, u32:32), false);
+    assert_eq(buffer_has_at_least(buffer, u32:33), false);
+
+    let buffer = Buffer { contents: u32:0, length: u32:16 };
+    assert_eq(buffer_has_at_least(buffer, u32:0), true);
+    assert_eq(buffer_has_at_least(buffer, u32:16), true);
+    assert_eq(buffer_has_at_least(buffer, u32:32), false);
+    assert_eq(buffer_has_at_least(buffer, u32:33), false);
+
+    let buffer = Buffer { contents: u32:0, length: u32:32 };
+    assert_eq(buffer_has_at_least(buffer, u32:0), true);
+    assert_eq(buffer_has_at_least(buffer, u32:16), true);
+    assert_eq(buffer_has_at_least(buffer, u32:32), true);
+    assert_eq(buffer_has_at_least(buffer, u32:33), false);
+}
+
+// Appends `data` to a `buffer`. Returns the updated buffer
+// It will fail if the buffer cannot fit the data. For calls that need
+// better error handling, check `buffer_append_checked`
+pub fn buffer_append<BSIZE: u32, DSIZE: u32> (buffer: Buffer<BSIZE>, data: bits[DSIZE]) -> Buffer<BSIZE> {
+    if buffer_can_fit(buffer, data) == false {
+        trace_fmt!("Not enough space in the buffer! {} + {} <= {}", buffer.length, DSIZE, BSIZE);
+        fail!("not_enough_space", ());
+    } else {
+        ()
+    };
+
+    Buffer {
+        contents: (data as bits[BSIZE] << buffer.length) | buffer.contents,
```

Comment:
I wonder if `data ++ buffer[0:buffer.length]` would be more readable here?

Reply:
Yes, it would. Unfortunately, It is impossible to use a variable as a slice index.
```
     } else {
         Buffer {
             contents: data ++ buffer.contents[0:buffer.length],
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^-----^ TypeInferenceError: Unable to resolve slice limit to a compile-time constant.
             length: DSIZE + buffer.length
         }
```

Reply:
that should work for `fixed_pop` (and is there a reason you need to support non fixed-sized pop?)

Reply:
given that #1168 only depends on `buffer_fixed_pop` maybe we could only include this variant for the time being (and simplify the implementation as initially suggested).